### PR TITLE
Add (advanced) preference for backup folder. 

### PR DESCRIPTION
**Pull request by Marr11317**
To get the old behavior, just set the setting `"application/backup/path"` to an empty string.

Its default value is `".mscbackup"`

EDIT: added a second commits that removes the old backup file

### CODE REVIEW

**Comment by Howard-C** [comment]
Hmm, this is probably too long, and in some cases this can cause some funny side effects. Why not `.mscbackup` like you suggested in #6016?

**Comment by Howard-C** [comment]
Or default to empty as suggested in https://t.me/musescoreeditorchat/72468.

**Comment by shoogle** [comment]
I'd rather use `.mscbackup` than make it configurable or users will change it and then forget that they did so. Or they might try changing it to an absolute path or something with `..` (parent directory). They might try setting it to `../../../../../home/me/.backup` or worse.

File: global/settings/types/preferencekeys.h [hunk 1]
```diff
@@ -52,6 +52,7 @@
 #define PREF_APP_STARTUP_TELEMETRY_ACCESS_REQUESTED         "application/startup/telemetry_access_requested"
 #define PREF_APP_TELEMETRY_ALLOWED                          "application/telemetry/allowed"
 #define PREF_APP_BACKUP_GENERATE_BACKUP                     "application/backup/generateBackup"
+#define PREF_APP_BACKUP_PATH                                "application/backup/path"
```
Comment (Spire42) [inline global/settings/types/preferencekeys.h:55]: This is a subfolder, not a path, so I suggest you change this to:

`#define PREF_APP_BACKUP_SUBFOLDER                           "application/backup/subfolder"`

File: libmscore/scorefile.cpp [hunk 2]
```diff
@@ -433,10 +436,11 @@ bool MasterScore::saveFile(bool generateBackup)
             // remove old backup file if exists
             // remove the backup file in the same dir as score (the traditional place) if exists
             //
-            QString backupDirString = info.path() + QString(QDir::separator()) + ".backup";
+            QString backupPath = preferences.getString(PREF_APP_BACKUP_PATH);
```
Comment (Spire42) [inline libmscore/scorefile.cpp:439]: For consistency, `backupPath` should be `backupSubdirString`.

Should also be `const QString`.

File: libmscore/scorefile.cpp [hunk 3]
```diff
@@ -433,10 +436,11 @@ bool MasterScore::saveFile(bool generateBackup)
...
+            QString backupDirString = info.path() + QString(QDir::separator()) + backupPath;
```
Comment (Spire42) [inline libmscore/scorefile.cpp:440]: This should also be `const QString`.

File: mscore/preferences.cpp [hunk 4]
```diff
@@ -123,6 +123,7 @@ void Preferences::init(bool storeInMemoryOnly)
             {PREF_APP_TELEMETRY_ALLOWED,                           new BoolPreference(false, false)},
             {PREF_APP_STARTUP_TELEMETRY_ACCESS_REQUESTED,          new StringPreference("", false)},
             {PREF_APP_BACKUP_GENERATE_BACKUP,                      new BoolPreference(true)},
+            {PREF_APP_BACKUP_PATH,                                 new StringPreference(".MuseScoreBackup")},
```
Comment (Spire42) [inline mscore/preferences.cpp:126]: I think this new feature should be opt-in, and so the default value of the `PREF_APP_BACKUP_SUBFOLDER` preference should be an empty string.

File: libmscore/scorefile.cpp [hunk 5]
```diff
@@ -36,6 +36,9 @@
 #include "stafftype.h"
 #include "sym.h"
 
+#include "mscore/preferences.h"
+
+
```
Comment (Spire42) [inline libmscore/scorefile.cpp:41]: Extra newline.

**Comment by Spire42** [comment]
> I'd rather use `.mscbackup` than make it configurable or users will change it and then forget that they did so. Or they might try changing it to an absolute path or something with `..` (parent directory). They might try setting it to `../../../../../home/me/.backup` or worse.

If this feature really must exist, I'd really rather it be configurable. We could always check to see if the value contains any `/` or `\` characters and then reject the value outright if so. In that case we would just fall back to the default, which I think should be changed to an empty string (to preserve the old behavior).

**Comment by shoogle** [comment]
As discussed on Telegram, another option instead of a folder is just to give the existing backup files a better name. Currently they end in `,` (comma) which is not very easy to see. Most text editors use `~` (tilde) which makes it easier to tell them apart from regular files.

**Comment by Marr11317** [comment]
> I'd rather use `.mscbackup` than make it configurable or users will change it and then forget that they did so. Or they might try changing it to an absolute path or something with `..` (parent directory). They might try setting it to `../../../../../home/me/.backup` or worse.

Is an advanced preference. it should really not be changed, be in case of some conflicts or whatever, it could be doable.

**Comment by Marr11317** [comment]
> As discussed on Telegram, another option instead of a folder is just to give the existing backup files a better name. Currently they end in `,` (comma) which is not very easy to see. Most text editors use `~` (tilde) which makes it easier to tell them apart from regular files.

That's an easy fix. But I have no idea how these are the backup files are opened, and if a comma is required on opening. @Howard-C you must know better than I.

**Comment by Howard-C** [comment]
@Marr11317 Yes, the file opening system recognizes `.mscz,` and `.mscx,` suffixes to determine backup file type. There's also code in `mscore/file.cpp` which deals with the `.` prefix too. So if either of these is changed, it affects other code.

**Comment by Marr11317** [comment]
> @Marr11317 Yes, the file opening system recognizes `.mscz,` and `.mscx,` suffixes to determine backup file type. There's also code in `mscore/file.cpp` which deals with the `.` prefix too. So if either of these is changed, it affects other code.

That's unfortunate. I'll leave the comma then.

File: mscore/preferences.cpp (hunk 4 above)
Comment (Marr11317) [reply mscore/preferences.cpp]: I personally don't see why backup files by not being grouped in a backup dir. I have been annoyed by all these backups stacking up.

**Comment by MarcSabatella** [comment]
FWIW, it's kind of a bug don't care for me - I was fine with the backup files going into he same folder,r I'm fine with them going into a hard-coded ",backup" subfolder (if other apps also use the same folder I don't see how this creates a problem, as surely they aren't trying to put files of the same name there), I'm fine with there being a way to configure the location, I'm fine with changing the name, etc.  Although I'd note changing the name won't satisfying those who complain about seeing them at all, and it will still result in people mistaking them for real MuseScoere files.  Particular on Windows and macOS, where many people simply don't understand the concept of file extensions at all because the OS often tries to hide them from you.

**Comment by Marr11317** [comment]
Maybe adding a 'backup' to the file name would help confused users?

**Comment by Marr11317** [comment]
The problem with changing now where backup files are stored is that there will now be two copies of backup files. One inside the subdirectory (the *new* backup), and one right beside the original file (the *old* backup). 
For now, I'll put this PR back to how it was on 3.4, because I think it is safer.

File: libmscore/scorefile.cpp [hunk 6]
```diff
@@ -433,25 +435,26 @@ bool MasterScore::saveFile(bool generateBackup)
...
 #ifdef Q_OS_WIN
-                  QString backupDirNativePath = QDir::toNativeSeparators(backupDirString);
+                  const QString backupDirNativePath = QDir::toNativeSeparators(backupDirString);
 #if (defined (_MSCVER) || defined (_MSC_VER))
    #if (defined (UNICODE))
                   SetFileAttributes((LPCTSTR)backupDirNativePath.unicode(), FILE_ATTRIBUTE_HIDDEN);
    #else
                   // Use byte-based Windows function
                   SetFileAttributes((LPCTSTR)backupDirNativePath.toLocal8Bit(), FILE_ATTRIBUTE_HIDDEN);
-   #endif
```
Comment (Jojo-Schmitz) [inline libmscore/scorefile.cpp:448]: If `unnesting` this, do it for the if and the else too. 
One portability trick BTW is to have the # in the first column, and spaces after that for indentation.

File: mscore/preferences.cpp [hunk 7]
```diff
@@ -123,6 +123,7 @@ void Preferences::init(bool storeInMemoryOnly)
             {PREF_APP_TELEMETRY_ALLOWED,                           new BoolPreference(false, false)},
             {PREF_APP_STARTUP_TELEMETRY_ACCESS_REQUESTED,          new StringPreference("", false)},
             {PREF_APP_BACKUP_GENERATE_BACKUP,                      new BoolPreference(true)},
+            {PREF_APP_BACKUP_SUBFOLDER,                            new StringPreference("")},
```
Comment (Jojo-Schmitz) [inline mscore/preferences.cpp:126]: I disagree with the default being empty, `.backup`, `.mscbackup`, whatever, starting with a dot (to make it hidden on Linux and Mac), but not back to the clutter we had before.

**Comment by jeetee** [comment]
Changing the filename of backup files might help in differentiating them better, but doesn't address the visual clutter of the directory listing complaint we were trying to address.

To that end, in my experience ~ is often used for temporary files, which indeed are mostly put directly in the same directory but are also erased on close.

The backup files from musescore aren't temporary. And I see little value in allowing customization of a subdirectory name. I do agree that `.backup` might seem to generic, so a fixed change to `.mscbackup` could make sense.

Their are many counterarguments for making it a setting; amongst which the raised concerns about relative paths. But most importantly those that come to the support forums and have changed that directory string. Then forgot about that. Then complain the backup scores aren't where we've told them to look.

**Comment by shoogle** [comment]
> (if other apps also use the same folder I don't see how this creates a problem, as surely they aren't trying to put files of the same name there)

It doesn't have to be a folder. The conflict also applies if there is a file called `.backup`.

**Comment by jeetee** [comment]
> > (if other apps also use the same folder I don't see how this creates a problem, as surely they aren't trying to put files of the same name there)
> 
> It doesn't have to be a folder. The conflict also applies if there is a file called `.backup`.

Ah yes, that seems a more plausible scenario indeed. So using `.mscbackup` would likely remove that conflict, agreed?

**Comment by Marr11317** [comment]
> Their are many counterarguments for making it a setting; amongst which the raised concerns about relative paths. 

The main reason I set it as a setting in the first place was to allow both keeping the old and the new behavior.

>But most importantly those that come to the support forums and have changed that directory string. Then forgot about that. Then complain the backup scores aren't where we've told them to look.

The relative path concerns doesn't seem to much of a problem since:
 1) It's an *advanced* preference. There is far more dangerous settings in the advanced tab.
 2) Users that use a relative path like `../../../home/me/desktop/.mscbackup` are either programmers or they know a bit about computers. They know what they're doing.

**Comment by Jojo-Schmitz** [comment]
> 2. Users that use a relative path like `../../../home/me/desktop/.mscbackup` are either programmers or they know a bit about computers. They know what they're doing.

Hopefully. But if not, it's all their fault anyway ;-)

**Comment by Marr11317** [comment]
So basically, the only concern left is about all the backup files that would've been created before 3.5

**Comment by jeetee** [comment]
> So basically, the only concern left is about all the backup files that would've been created before 3.5

Which is a non-issue to me.

**Comment by Marr11317** [comment]
switched back to `".mscbackup"` as default

**Comment by Marr11317** [comment]
> > So basically, the only concern left is about all the backup files that would've been created before 3.5
> 
> Which is a non-issue to me.

Anyway, I just took care of that. It's just an idea, and if it turns out not to be wanted, simply don't merge the second commit.

**Comment by Jojo-Schmitz** [comment]
I like the idea

**Comment by jeetee** [comment]
I like the idea as well, but wonder what happens when opening a backup file in MuseScore. Is it already "imported" at that stage and thus decoupled from the file itself?

**Comment by Jojo-Schmitz** [comment]
You mean imported but not yet saved? 
Worst thing, depending on operating system and file system is that the delete fails, silently.
Or that you can't import it afresh.
IMHO neither is something to worry about.

**Comment by shoogle** [comment]
What about files inside the backup folder? Presumably they don't need to be hidden or have the comma extension.

**Comment by Marr11317** [comment]
> What about files inside the backup folder? Presumably they don't need to be hidden or have the comma extension.

And what problem could be attached with not hiding them?

**Comment by shoogle** [comment]
>> What about files inside the backup folder? Presumably they don't need to be hidden or have the comma extension.
>
> And what problem could be attached to hiding them?

No big "problem", but it means an extra step in renaming the file and removing the hidden flag (on Windows) when restoring from backup.

**Comment by jeetee** [comment]
> No big "problem", but it means an extra step in renaming the file and removing the hidden flag (on Windows) when restoring from backup.

The main though (or at least mine :) ) there is that the user wouldn't need to do this manually anymore. To access the hidden .mscbackup directory they already have to enable viewing of hidden files & folders at that moment. So opening that folder will already show that listing.
But I don't see the need to hide them either, so I'm quite neutral on this part.

Once they've located their backup file, I'd expect them to double click it. *Here the extension needs to persist* as it allows MuseScore to treat this score as an import.
We could even go as far as then suggesting the parent folder as the "Save As..." location if the opened file resided in a .mscbackup folder?

**Comment by shoogle** [comment]
> Once they've located their backup file, I'd expect them to double click it. *Here the extension needs to persist* as it allows MuseScore to treat this score as an import.
> We could even go as far as then suggesting the parent folder as the "Save As..." location if the opened file resided in a .mscbackup folder?

So check extension on import and folder name on export. Why not check the folder name on both? That way you don't need an extension.

**Comment by Marr11317** [comment]
> So check extension on import and folder name on export. Why not check the folder name on both? That way you don't need an extension.

Because it won't work for backups that do not reside where musescore expects them to.

**Comment by shoogle** [comment]
Backups outside the special backup folder would still be hidden and have the extension.

**Comment by Marr11317** [comment]
There might be a way we make this work, but it just seems so much simpler to me to just keep the current behaviour, that is everything is hidden, including files **and** directories, and keep the trailing comma.

File: libmscore/scorefile.cpp [hunk 8]
```diff
@@ -451,7 +454,18 @@ bool MasterScore::saveFile(bool generateBackup)
...
+
+            // backup files prior to 3.5 were saved in the same directory as the file itself.
+            // remove these old backup files if needed
+            if (dir != backupDir && dir.exists(backupName)) {
+                  if (!dir.remove(backupName)) {
+//                      if (!MScore::noGui)
+//                            QMessageBox::critical(0, QObject::tr("Save File"),
+//                               tr("Removing old backup file %1 failed").arg(backupName));
+                        }
+                  }
```
Comment (Howard-C) [inline libmscore/scorefile.cpp:467]: There's already code for that several lines below. Just modify the conditional to `if (dir != backupDir && dir.exists(backupName))` would be good.

Comment (Marr11317) [reply libmscore/scorefile.cpp:467]: done

File: libmscore/scorefile.cpp [hunk 9]
```diff
@@ -448,21 +448,24 @@ bool MasterScore::saveFile(bool generateBackup)
...
                   SetFileAttributes((LPCTSTR)backupDirNativePath.toLocal8Bit(), FILE_ATTRIBUTE_HIDDEN);
-#endif
+   #endif
 #else
                   SetFileAttributes((LPCTSTR)backupDirNativePath.toLocal8Bit(), FILE_ATTRIBUTE_HIDDEN);
 #endif
 #endif
                   }
             const QString backupName = QString(".") + info.fileName() + QString(",");
+
```
Comment (Jojo-Schmitz) [inline libmscore/scorefile.cpp:458]: If an empty line, then rather above

**Pull request merged**